### PR TITLE
Arm thumb ADR instruction reference pointer

### PR DESCRIPTION
**Pull request by jg10**
I have added a word aligned ARM thumb mode ADR instruction relative offset pointer to help resolve strings that might be referenced in this way. I was analysing an ARM firmware that referenced all strings with the ADR instruction which uses a word aligned offset from the instruction pointer and radare2 was not correctly resolving it. This patch fixes that and will allow radare2 to correctly resolve simple strings that are referenced by the ADR instruction which makes static analysis much easier and also allows proper cross references too. The example below which I'll also be submitting separately as a radare2-regressions test demonstrates a string being referenced by the ADR instruction with this code change.

```bash
[0x00000000]> e asm.bits=16
[0000:0000]> e asm.arch=arm
[0x00000000]> wx 10b5 01a0 00bf 00bf 52616461726532207465737420737472696e6700
[0x00000000]> pd 10
            0x00000000      10b5           push {r4, lr}
            0x00000002      01a0           adr r0, 4                   ; 0x8 ; "Radare2 test string"
            0x00000004      00bf           nop
            0x00000006      00bf           nop
            0x00000008      5261           str r2, [r2, 0x14]
            0x0000000a      6461           str r4, [r4, 0x14]
            0x0000000c      7265           str r2, [r6, 0x54]
            0x0000000e      3220           movs r0, 0x32               ; '2'
            0x00000010      7465           str r4, [r6, 0x54]
            0x00000012      7374           strb r3, [r6, 0x11]
```

If a string has a \r and/or \n in it though it won't show up here and I will submit a another pull request to also display them however I will do it separately as I don't know if they will have flow on effects as the code changes are in libr/core/disasm.c and libr/util/str.c and could effect other analysis.

### CODE REVIEW

File: libr/anal/p/anal_arm_cs.c [hunk 1]
```diff
@@ -1801,6 +1801,12 @@ jmp $$ + 4 + ( [delta] * 2 )
 			op->fail = addr + op->size;
 		}
 		break;
+	case ARM_INS_ADR:
+		op->type = R_ANAL_OP_TYPE_LEA;
+		// Set the pointer address and align it
+		op->ptr = IMM(1) + addr + 4 - (addr%4);
```
Comment (alvarofe) [inline libr/anal/p/anal_arm_cs.c]: Why addr % 4? if in thumb mode shouldn't be 2?

Comment (jg10) [inline libr/anal/p/anal_arm_cs.c]: In the 'Thumb ADR 16-bit encoding' line of Table 10-4 Register-relative offsets at  http://www.keil.com/support/man/docs/armasm/armasm_dom1361289862147.htm#/dom1361289862147__adr_reg_mult_fn it has a footnote that the offset range must be a multiple of 4.

I also found in the firmware I was analysing that all strings were aligned to addresses ending in 0,4,8,c so by subtracting addr%4 it always aligned and all strings resolved correctly.

Comment (jg10) [inline libr/anal/p/anal_arm_cs.c]: Sorry I think that should have been this link for the PC-relative ADR instruction and Table 10-3 PC-relative offsets http://www.keil.com/support/man/docs/armasm/armasm_dom1361358903373.htm

**Comment by radare** [comment]
i dont think its a good idea to display newlines in strings inside the disassembly, it may mess up the disasm pretty quickly

**Comment by radare** [comment]
can u send a PR for r2r with that test case?

**Comment by jg10** [comment]
I just sent the PR for r2r (#727). I also completely agree that it's not a good idea to display newlines in the disassembly (I saw that first hand while I was developing the changes). I will propose a change that basically adds a further check for \n and \r to r_str_is_printable in str.c and then uses r_str_filter to replace the \n and \r with dots in disasm.c. So "test string\r\n" will print in disassembly as "test string.."

**Pull request merged**